src/astar.rs

```diff
@@ -0,0 +1,171 @@
+use std::collections::{
+    HashMap,
+    BinaryHeap,
+};
+use std::collections::hash_map::Entry::{
+    Occupied,
+    Vacant,
+};
+
+use std::hash::Hash;
+
+use scored::MinScored;
+use super::visit::{
+    EdgeRef,
+    GraphBase,
+    IntoEdges,
+    VisitMap,
+    Visitable,
+};
+
+use algo::Measure;
+
+/// [Generic] A* shortest path algorithm.
+///
+/// Compute the length of the shortest path from `start` to `finish`.
+///
+/// `finish` is implicitly given via the `is_goal` callback, which should return `true` if the
+/// given node is the finish node.
+///
+/// The function `edge_cost` should return the cost for a particular edge, which is used to compute
+/// path costs.
+///
+/// The function `estimate_cost` should return the estimated cost to the finish for a particular
+/// node, also used to compute path costs.
```

Comment:
This sentence needs clarification. I think I can do that, but I wonder what you think too. I think the "also used to compute path costs" is confusing and should be removed. Instead I'd expect it to clarify us what the requirements of the heuristic are, and as wikipedia says:

> For the algorithm to find the actual shortest path, the heuristic function must be admissible, meaning that it never overestimates the actual cost to get to the nearest goal node

Reply:
I agree with you 100% on this. I haven't checked the docstring for quite some time, I need to give it another read to see what needs to be rewritten.

Reply:
I don't want to give you more to do though!! :-) I have a comment on tests so I can't stop idly trickling in comments.